[API-1048] Make intellisense and type hints available on blocking proxies

mdumandag opened the pull request:
This PR attempts to provide type hints and intellisense for the
blocking proxies by introducing blocking versions of the proxies
that implement the same interface with different return types.

With this approach, we don't have to duplicate the documentation,
and return the blocking subclasses directly to the user.

Actually, the blocking proxies are not really subclasses, but very
thin wrappers around the public APIs of the proxies. They look like
subclasses to make the documentation of the actual proxies available
to them.

This makes mypy complain, but it is possible to ignore such
warnings about return type mismatches between child and parent classes.
So, we could work around that problem once we start using mypy.

Also, we have modified the return types and documentation of transactional
proxies. It wasn't possible to use them in non-blocking mode so I changed
the return types and documentation to reflect that.

yuce commented on hazelcast/cluster.py:
```diff
@@ -62,49 +63,63 @@ class ClusterService:
-        It will be notified when a member is added to cluster or removed from cluster.
-        There is no check for duplicate registrations, so if you register the listener
-        twice, it will get events twice.
+        It will be notified when a member is added to cluster or removed from
```
How about: "... added to the cluster or removed from the cluster ..."?

yuce commented on hazelcast/cluster.py:
```diff
@@ -31,10 +32,10 @@ class ClientInfo:
-        self.address = address
-        self.name = name
-        self.labels = labels
+        self.uuid: uuid.UUID = client_uuid
```
How about annotating the `client_uuid` argument instead of `self.uuid`?

yuce commented on hazelcast/core.py:
```diff
@@ -412,23 +415,23 @@ class HazelcastJsonValue:
     """
 
-    def __init__(self, value):
+    def __init__(self, value: typing.Union[str, typing.Any]):
```
Can `typing.Union[str, typing.Any]` be simplified to `typing.Any`? or just leave out the annotation altogether?

yuce commented on hazelcast/future.py:
```diff
@@ -100,13 +102,13 @@ def exception(self):
         return self._traceback
 
-    def add_done_callback(self, callback):
+    def add_done_callback(self, callback: typing.Callable[["Future"], None]) -> None:
```
Does it make sense to move `typing.Callable["Future"]` to `hazelcast.typing`?

yuce commented on hazelcast/future.py:
```diff
@@ -127,14 +129,19 @@ def _invoke_cb(self, callback):
-    def continue_with(self, continuation_func, *args):
+    def continue_with(
+        self,
+        continuation_func: typing.Callable[..., typing.Union["Future", typing.Any]],
```
Can `typing.Union["Future", typing.Any]` be simplified to `typing.Any`?

yuce commented on hazelcast/predicate.py:
```diff
@@ -360,9 +362,10 @@ def __init__(self, predicate, page_size, comparator=None):
+        self.comparator: typing.Any = comparator
+        self.iteration_type: int = IterationType.ENTRY
+        # List of pairs: (nearest page, (anchor key, anchor value))
+        self.anchor_list: typing.List[typing.Tuple[int, typing.Tuple[typing.Any, typing.Any]]] = []
```
`typing.Tuple[typing.Any, typing.Any]` is equivalent to `tuple` or `typing.Tuple`, you may consider using the shorter alternatives.

mdumandag replied:
I did this on purpose, to make it more clear that it contains two elements of type any. If it is not a problem for you, I would like to keep it like that.

I also used typing.Any or None return type in places to make sure that I did not miss anything.

mdumandag replied:
BTW, we can't make use of `tuple` or `list` as they are not available in Python 3.6

yuce commented on hazelcast/predicate.py:
```diff
@@ -586,7 +589,7 @@ def not_(predicate):
 
 
-def between(attribute, from_, to):
+def between(attribute: str, from_: typing.Any, to: typing.Any) -> Predicate:
```
I think `typing.Any`s can be safely left out. There are more throughout the code.

mdumandag replied:
See my comment above. It is done on purpose

yuce commented on hazelcast/projection.py:
```diff
@@ -49,9 +51,8 @@ def get_class_id(self):
-class _MultiAttributeProjection(_AbstractProjection):
-    def __init__(self, attribute_paths):
-        # type: (list[str]) -> None
+class _MultiAttributeProjection(_AbstractProjection[typing.List[typing.Any]]):
```
`typing.List[typing.Any]` can be simplified to `typing.List` or just `list`. There are more.

yuce reviewed (commented):
Here are a few ideas. Continuing to review.

yuce commented on hazelcast/proxy/cp/fenced_lock.py:
```diff
@@ -117,15 +119,15 @@ def lock(self):
         return self._do_lock(current_thread_id, invocation_uuid)
 
-    def try_lock(self, timeout=0):
+    def try_lock(self, timeout: Numeric = 0) -> Future[int]:
```
I think `timeout` should be just `float`.

yuce commented on hazelcast/proxy/flake_id_generator.py:
```diff
@@ -10,26 +10,30 @@
+    """A cluster-wide unique ID generator. Generated IDs are int values and
+    are k-ordered (roughly ordered). IDs are in the range from 0 to 2^63 - 1.
+
+    The IDs contain timestamp component and a node ID component, which is
```
"... contain a timestamp ..."

yuce commented on hazelcast/proxy/flake_id_generator.py:
```diff
@@ -46,24 +50,28 @@ def __init__(self, service_name, name, context):
-        This method goes to a random member and gets a batch of IDs, which will then be returned
-        locally for limited time. The pre-fetch size and the validity time can be configured.
+        This method goes to a random member and gets a batch of IDs, which will
+        then be returned locally for limited time. The pre-fetch size and the
```
"... for a limited time."

mdumandag commented on hazelcast/proxy/cp/atomic_long.py:
```diff
@@ -52,116 +55,116 @@ def compare_and_set(self, expect, update):
-            ``Portable`` that has a counterpart that implements the
-            ``com.hazelcast.core.IFunction`` interface registered on the server-side with
-            the actual implementation of the function to be applied.
+            ``function`` must be an instance Hazelcast serializable type that
+            has a counterpart that implements the
+            ``com.hazelcast.core.IFunction`` interface registered on the
+            server-side with the actual implementation of the function to be
+            applied.
```
I will change it to 
```
            ``function`` must be an instance of Hazelcast serializable type.
            It must have a counterpart registered in the server-side that
            implements the ``com.hazelcast.core.IFunction`` interface with
            the actual logic of the function to be applied.
```

WDYT?

mdumandag commented on hazelcast/proxy/cp/atomic_reference.py:
```diff
@@ -6,35 +9,40 @@
-      calculate some information or need a subset of fields, you can use the `apply()`
-      method. With the `apply()` method, the whole object does not need to be sent over
-      the line; only the information that is relevant is sent.
+    The following are some considerations you need to know when you use
+    AtomicReference:
```
I feel like the current version is better

mdumandag commented:
@yuce I have gone over your comments and fixed most of them. I put comments on the ones that I didn't apply. 

BTW, can you do your further reviews in #518, as it is the base PR that we need to merge? I have applied your comments to there and just rebased here

utku-caglayan reviewed (approved):
LGTM!

mdumandag commented:
Thanks for the reviews guys!
